content/post/graceful-k8s-delpoyments/index.md

```diff
@@ -0,0 +1,205 @@
+---
+title: "Zero-Downtime Rolling Deployments in Kubernetes"
+date: 2021-07-1500:00:00+09:00
+hero: /images/rolling-waves.jpg
+excerpt: |
+  Kubernetes enables us to operate robust systems at scale. However, it is not perfect and in edge cases k8s may behave in unexpected ways.
+authors:
+  - Manish Yadav
+  - Oscar Garcia
+  - Nemo
+  - Boaz Yaniv
+draft: true
+license: cc-by
+---
+
+*[Photo by Kammeran Gonzalez-Keola](https://www.pexels.com/@kammeran-gonzalez-keola-3137381) used under [Pexels License](https://www.pexels.com/license/)*
+
+At Rakuten we default to Kubernetes to run our services, and k8s enables us to operate robust systems at scale. However, it is not perfect and in edge cases k8s may behave in unexpected ways. For example, during rolling deployments we expect no request failures - a reasonable expectation, right? But when we ran functional tests during a deployment we notice requests failing - to our surprise. After researching this for a while we learned that...
+
+> K8s will send traffic to terminating pods.
+
+Even after starting the pod shutdown process, which includes sending a `TERM` signal to pod, k8s still sends new requests to the pod. Because there is no orchestration between k8s sending `TERM` signal and removing the pod from its service endpoint list. Those 2 operations can happen in any order, possibly with delay in between them. Yikes 😬.
```

Comment:
> Even after starting the pod shutdown process, which includes sending a `TERM` signal to pod, k8s still sends new requests to the pod.

Is this correct? [Google says traffic stops as soon as the pod receives `SIGTERM`](https://cloud.google.com/blog/products/containers-kubernetes/kubernetes-best-practices-terminating-with-grace#:~:text=At%20this%20point%2C%20the%20pod%20stops%20getting%20new%20traffic.%20Containers%20running%20in%20the%20pod%20will%20not%20be%20affected.), before any `preStop` hook gets called.

Reply:
Confirmed that Google is lying 😂